app/bundles/CampaignBundle/Executioner/Helper/EventRedirectionHelper.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Mautic\CampaignBundle\Executioner\Helper;
+
+use Doctrine\Common\Collections\Collection;
+use Mautic\CampaignBundle\Entity\Event;
+use Psr\Log\LoggerInterface;
+
+/**
+ * Helper class for handling event redirection.
+ * This helper centralizes logic for finding and following chains of redirected events.
+ * It also provides methods to handle the replacement of deleted events with their redirects.
+ */
+class EventRedirectionHelper
+{
+    private const MAX_DEPTH = 20;
+
+    public function __construct(
+        private LoggerInterface $logger,
+    ) {
+    }
+
+    /**
+     * Handle redirection for a deleted event and replace it in the event's collection.
+     * If a redirection is found, returns the redirected event; otherwise returns the original event.
+     *
+     * @param Event       $event  The event to check for redirection
+     * @param ?Collection $events The collection containing the event
+     * @param mixed       $key    The key of the event in the collection
```

Comment:
> The key can either be an [int](https://www.php.net/manual/en/language.types.integer.php) or a [string](https://www.php.net/manual/en/language.types.string.php).

Source: https://www.php.net/manual/en/language.types.array.php

The same applies for Doctrine entities. I can see that you need it also nullable for some reason.
```suggestion
     * @param int|string|null       $key    The key of the event in the collection
```

Reply:
Fixed.